### ISSUE TITLE
use cryosparc output as relion input

### ISSUE THREAD

**Issue by rmarabini** [scipion-em/scipion-em-relion#371]
Hi @pconesa, @fonsecareyna82 and @azazellochg,

Let me explain a problem that appears when you try to use the output of "cryosparc 3D-non uniform refinement" (and I guess any other refinement) as input for "relion 3D autorefine" (and I guess any 3D refine).  That is, when I further refine using relion (with the option "consider previous alignment" set to True) a setOfParticles created by cryosparc.  In my test case the particle has I2 symmetry. 

Relion and cryosparc handle symmetry in a different way, Relion takes advantage of the symmetry and, for I2 symmetry, only searches in 1/60 of all possible projection directions while cryosparc covers all the sphere. In this way if you plot the projection directions in the Relion case you see that all projection direction are concentrated in a small triangle defined by 2 5fold symmetry axis and 1 3fold symmetry axis while Cryosparc shows points in all directions. 

So, what happens when relion reads the cryosparc data? Since relion only has information for the projection directions inside the non redundant part of the projection sphere, it will take the  cryospac projection directions and  assign then the value of the closest relion-valid projection direction (I hope I has been able to explain myself). In general, the closest valid projection direction is not close to the original projection and the  reconstruction is a shapeless sphere. 

I have created a script that takes as input the cryosparc setOfParticles and the symmetry and produces a setOfPArticles that Relion likes. The question now is how to incorporate this script in Scipion. The first option, that I dislike a lot, is create a protocol called "adapt setof particles for Scipion", The second option would be to decide what convention Scipion is going to follow and adapt the protocols to it. In this way if Scipion follows Relion convention then cryosparc needs to rearrange the setOfPraticles projection directions before saving them. On the other hand if Scipion follows cryosparc convention then relion protocols need to preprocess the projection direction before using them. 

If the decision was mine I would choose relion conventions as scipion conventions and modify the output of cryosparc. 
what do you think?

**Comment by pconesa** [comment]
Thanks @rmarabini I got it, actually some time ago I faced this issue when trying to add an "angular distribution plot" for CS and found CS does not care about symmetry (only when processing)

My first guess would be to "migrate CS angular distribution to the non redundant space" inside CS refinement protocols. This way all will be fine for relion and the viewers.

Now, I'm not sure if CS would like this approach when its input is like this (theoretically it should work, right?)

**Comment by rmarabini** [comment]
Hi @pconesa,

   cryosparc is very introvert,  you can not provide "cryosparc 3D-refinement-XXXX"  with a initial set of projection directions so the actual angular distribution is not important. Furthermore, cryosparc (or at least cryosparc though scipion) does not have a plain reconstruction algorithm so I do not think there is a situation in which the initial angular distribution matters for cryosparc.

> Thanks @rmarabini I got it, actually some time ago I faced this issue when trying to add an "angular distribution plot" for CS and found CS does not care about symmetry (only when processing)
> 
> My first guess would be to "migrate CS angular distribution to the non redundant space" inside CS refinement protocols. This way all will be fine for relion and the viewers.
> 
> Now, I'm not sure if CS would like this approach when its input is like this (theoretically it should work, right?)

**Comment by pconesa** [comment]
You're right, I think there is a reconstruction method that might need it.

**Comment by azazellochg** [comment]
https://discuss.cryosparc.com/t/how-does-cryosparc-treat-symmetry/2787/2
pyem has star2bild.py script which has a --sym argument that applies symmetry to the angles. If you want to force the angles into one part of the rotation sphere, it can be used at the end of createoutputstep in the CS plugin

**Comment by rmarabini** [comment]
Thanks for the information @azazellochg  ;-). I am in the process of writing a similar script. There is an extra problem here, what cryosparc defines as "non-reduntant part of the rotation-sphere" sometimes is different from what relion (and other packages) defines as "non-reduntant part of the rotation-sphere". So we need to define a standard for Scipion an convert to/from this standard when needed.

**Comment by pconesa** [comment]
@rmarabini , could we meet to address this together?

**Comment by rmarabini** [comment]
@pconesa sure. Remember that I wrote a couple of methods that  can handle this task. These methods are in symmetry.py  (scipion-em). For example

def moveParticlesInsideUnitCell(setIN, setOUT, sym=cts.SYM_CYCLIC, n=1):
    """ Move particles projection direction inside the unit cell associated
    to symmetry sym.
    This function (1) gets the symmetry matrices and (2) applies them to
         each projection direction until they are inside the unit cell.
    """

Whenever possible symmetries are defined as chimera uscf does.
Note that relion and cryo-spark are not fully compatible in the sense that support a subset of symmetries. For example,  cryosparc defines Dn as Cn around z + C2 around Y, while relion and xmipp, and maybe others, define it as Cn around z + C2 around X.  Therefore here we need to decide what todo.  Possibilities:

1) scipion stores both  options as is up to the developers to convert it properly at the star of each protocol
2) Scipion only stores one of the options and is up to the developer of a particular package to convert the metadata at the end of each protocol.

 For this task you may use you may use

        def coordinateSystemTransform(self, origSym, targetSym):

which is also in symmetry.py

Examples of use are in pwem/tests/data/test_symmetry.py and in pwem/protocols/protocol_projection_edit.py
Finally scipion symmetry definitions are at URL: https://scipion-em.github.io/docs/release-3.0.0/docs/developer/symmetries/symmetries.html?highlight=symmetry

good luck

    Roberto

**Comment by pconesa** [comment]
I think this is addressed now by moving particles to our single unit cell in cryosparc. I think this ca be closed
